### PR TITLE
Plugin framework

### LINKED ISSUE #2113: Plugins: types and context
## Is your feature request related to a problem? Please describe 👀
Part of epic #2084 
Create the types and a context for the plugins within common & host packages.

## Describe the solution you'd like 🎁

Some types, to be exposed by the common package:

```
export enum PluginArea {
  AppBar,
  Toolbar,
}

export enum PluginType {
  InboundShipment,
  InternalOrder,
  OutboundShipment,
  Requisition,
  Stocktake,
}

export interface Plugin<T> {
  area: PluginArea;
  component?: () => React.JSX.Element;
  data?: T;
  name: string;
  path: string;
  type: PluginType;
}

export interface PluginDefinition {
  config: string;
  name: string;
}
```

Then a `PluginProvider` in the host package which can:

- fetch the list of plugins, parse the config, and, for each element:
- store in an array of plugin objects with area, component, name, type, path

Create a `PluginLoader` component which can render the plugin, using `React.Lazy` and `React.Suspense` with an `ErrorBoundary` to keep things safe and isolated. 

Create a hook to fetch applicable plugins for a given area and type

### Moneyworks Jobcode 🧰

`OMSPLUG`

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2113 

# 👩🏻‍💻 What does this PR do? 
 
Creates the plugin framework - adds types for the plugins, and a Provider / Context which is populated on app load with the available plugins. 

Each plugin exposes an `area` and `type` and these are used to determine if a plugin is required to be shown in a particular area.

The host `PluginsProvider` fetches plugins using the `plugins` query and stores the array of plugins in the `PluginProvider`. Client areas can then call the `usePlugins` hook to fetch any plugins which relate to that specific area and type, or the `usePluginElements` hook to fetch the JSX.Elements for each applicable plugin

For example - if a plugin exposes a button which should be added to the `AppBarButtons` of the InboundShipment DetailView, it can call the hook:

```
  const pluginButtons = usePluginElements({
    area: PluginArea.Toolbar,
    type: PluginType.InboundShipment,
    data,
  });
```

and simply render the results in the `AppBarButtonsPortal` along with the standard buttons:

```
        {pluginButtons}
```

# 🧪 How has/should this change been tested? 

The [example-01](https://github.com/openmsupply/open-msupply/tree/feature/front-end-plugins-example-01) is the easiest way to test the full solution - just start the server and run `yarn start-local` for the client, and then navigate to an inbound shipment detail page. There should be a new button (View) in the top right menu:

Screenshot 2023-08-29 at 5 09 36 PM

which opens a modal:

Screenshot 2023-08-29 at 5 09 46 PM

Alternatively - you'd have to create a plugin, am in the process of writing instructions for that. The remainder of testing is a general site test to see if anything is broken.

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

File: client/packages/config/src/config.ts [hunk 1]
```diff
@@ -1,31 +1,39 @@
...
 const { port, hostname, protocol } = window.location;
 
 const defaultDevelopmentApiHost = `${protocol}//${hostname}:8000`;
 const productionApiHost = `${protocol}//${hostname}:${port}`;
 
-const developmentApiHost = 
+const developmentApiHost =
   (typeof API_HOST !== 'undefined' && API_HOST) || defaultDevelopmentApiHost;
-
+console.info('hello', PLUGIN_URL);
```
Comment (clemens-msupply) [inline client/packages/config/src/config.ts:26]: debug output?

File: client/packages/config/src/config.ts [hunk 2]
```diff
@@ -1,31 +1,39 @@
...
 
-// For development, API server and front end are launched seperately on different ports and possible different IPs
+// For development, API server and front end are launched separately on different ports and possible different IPs
 // by default we assume development API server is launched on the same domain/ip and on port 8000. We can overwrite this
 // with API_HOST which is available through webpack.DefinePlugin (i.e. webpack server --env API_PORT=800 --env API_IP 'localhost')
 
-// Important to note, if we overwrite API_HOST in development, we should use ip/domain that is known outsideof localhost
-// because web app in development mode may be accessed by clients on different machine (i.e. when debugin Android app)
+// When developing plugins, you can set the env variable PLUGIN_URL to the url of the plugin server to enable hot reloading
+// e.g. `webpack-cli serve --port 3000 --env PLUGIN_URL=http://localhost:3000`,
```
Comment (clemens-msupply) [inline client/packages/config/src/config.ts:13]: What is the plugin server? the remote server?

File: client/packages/common/src/plugins/utils.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  async (): Promise<PluginModule> => {
+    try {
+      // Check if this plugin has already been loaded
+      if (!(plugin in window)) {
+        // Initializes the shared scope. Fills it with known provided modules from this build and all remotes
+        await __webpack_init_sharing__(scope);
+        // Fetch the plugin app
+        const fetchedContainer = await fetchPlugin(`${url}`, plugin);
+        // Initialize the plugin app
+        await fetchedContainer.init(__webpack_share_scopes__[scope]);
```
Comment (clemens-msupply) [inline client/packages/common/src/plugins/utils.tsx:61]: do you know how this works? e.g. where does the containter get the init() method from?

**Review by clemens-msupply — COMMENTED** [review]
Looks very good! mainly questions..

I checkout out the example branch but can't see the new button, any idea why? Do I have to do something on the backend side as well?

Also not sure how it works to have the plugin in the frontend code and how that gets loaded... But I guess I wait for the docs :-)

File: client/packages/common/src/plugins/utils.tsx (hunk 3 above)
Comment (mark-prins) [reply client/packages/common/src/plugins/utils.tsx:61]: the webpack module federation is adding the `get` and `init` methods when bundling. 
this process is lifted from the docs and webpack examples

**Comment by mark-prins** [comment]
> I checkout out the example branch but can't see the new button, any idea why? Do I have to do something on the backend side as well?

ah yes. I shall have to start writing docs!
The server needs to have a folder for the plugin, under `app_data/plugins`
If you create a `ShippingStatus` folder and then run the build for the plugin `build-plugin` then the plugin build will compile and copy to the server folder.

The process is a little disjointed in that way: the server is enumerating all folders under the plugin directory to get a list of available plugins and the client is using that list to determine if a plugin is available.

So even if the plugin is running, via `yarn start-local` it won't be displayed unless the server includes it in the list of available plugins.

File: client/packages/common/src/intl/locales/en/common.json [hunk 4]
```diff
@@ -66,6 +66,8 @@
   "error.no-reports-available": "No reports available",
   "error.no-results": "Nothing here",
   "error.no-service-charges": "There are no service charges assigned to this store",
+  "error.plugin": "There was an error loading the plugin {{path}}",
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/common/src/intl/locales/en/common.json:69]: thanks, yes - that's better

File: client/packages/config/src/config.ts (hunk 2 above)
Comment (mark-prins) [reply client/packages/config/src/config.ts:13]: ah, it should be the local webpack dev server of the plugin app. I've updated the comment - thanks

**Review by clemens-msupply — APPROVED** [review]
Very cool, works nicely!

Noticed that the path could be shortened, e.g. just `"path": "ShippingStatus/plugin.json"`:

```json
{
    "data": {
        "plugins": [
            {
                "config": "{\n    \"name\": \"ShippingStatus\",\n    \"version\": \"1.0.0\",\n    \"components\": [{\n      \"area\": \"Toolbar\",\n      \"type\": \"InboundShipment\",\n      \"module\": \"Button\"\n    }],\n    \"dependencies\": {\n    \"omSupplyVersion\": \"8.0.0\"\n  }\n}",
                "name": "ShippingStatus",
                "path": "app_data/plugins/ShippingStatus/plugin.json"
            }
        ]
    }
}
```

**Comment by mark-prins** [comment]
> Noticed that the path could be shortened, e.g. just "path": "ShippingStatus/plugin.json":

thanks! yes, I've changed in the API PR, and will merge the changes into here if they are ok

**Review by andreievg — COMMENTED** [review]
Hi, that looks very good thanks !

I notice that potentially plugin is being loaded more then once ?

Screenshot 2023-09-05 at 4 51 37 PM

I was wondering about cataloging plugins by the categories, why we would need this ?

I was also interested in making this work with hot reload and adding type safety. Before I paused on the plugin research work, those two seemed like pretty hard requirements. Also I was hoping we would create a plugin system pattern that's easy to replicate and add it to areas on per need bases.

For type safety a type like this would work:

```typescript
type ReactElementResult = {/* some jsx type */};
type ReactComponent <T> = (_: T) => ReactElementResult;

type InvoiceNode = 'invoiceNodeThingsHere'
type RequisitionNode = 'requisitionNodeThingsHere'

type InboundButtonPlugin = {Component: ReactComponent<{invoiceNode: InvoiceNode}>, pluginType: 'InboundButtonPlugin'};
type OutboundButtonPlugin = {Component: ReactComponent<{requisitionNode: RequisitionNode}>, pluginType: 'OutboundButtonPlugin'};
type PluginTypes = InboundButtonPlugin | OutboundButtonPlugin;

// This is plugin definition, it's type safe for inputs
const inboundButtonPlugin: InboundButtonPlugin = {
    // Any other input type will error out
    Component: ({invoiceNode}) => { 
        return {}
    },
    pluginType: "InboundButtonPlugin" // Any other output will error out
}

const plugins: PluginTypes[] = [inboundButtonPlugin];

const getPlugins = <PT extends PluginTypes['pluginType']>(pluginType: PT): Extract<PluginTypes, {pluginType: PT}>[]  => {
    // This way can narrow down on the plugin type
    return plugins.filter((plugin): plugin is Extract<PluginTypes, {pluginType: PT}> => plugin.pluginType === pluginType)
}

// When plugin is used, can be type safe also
const inboundButtonPlugins = getPlugins('InboundButtonPlugin');

for (let plugin of inboundButtonPlugins) {
    plugin.Component({invoiceNode: "invoiceNodeThingsHere"});
    // Below line will not compile
    // plugin.Component({requisitionNode: "requisitionNodeThingsHere"});
}
```

For hot reload we can use lazy import() from some front end folder and import plugins directly.

I ended up making example of what I mean here: https://github.com/andreievg/plugins-example

Sorry, I know this is a bit of work but I was meaning to make a tutorial at some point about this anyways (including back end plugins), and seeing as this is an area of high importance, I though to spend a bit of time on this myself, and share what I was imagining.

**Comment by mark-prins** [comment]
> I was wondering about cataloging plugins by the categories, why we would need this ?

Not sure what you mean- can you elaborate?

> those two seemed like pretty hard requirements

I disagree about HMR and don't see that as a hard requirement. Besides which, hot reloading is working already in the examples I've done, although due to the convoluted loading nature, the whole page refreshes. Still, plugins are not intended as large applications and I don't expect that a hot reload is a big issue.

Re the type safety though: yes, that bothers me in what's been done so far. I don't like the approach of having specific plugin types though, having the types of available plugins hard coded and defined in code - I was trying to avoid that inflexibility. OTOH the plugins will need to have an implementation in the hosting app, so there is already a limitation imposed

**Comment by mark-prins** [comment]
> Hi

[from @clemens-msupply]

Good idea, think this would be cool! Alternatively, we could maintain multiple lists of plugin types and then have methods like getInboundShipmentPlugins(). At the plugin boundary we can only do limited plugin validation anyway and have to cast to the expected type at some point, i.e. trust that the plugin is implemented correctly. Anyway after this initial type cast we should use the correct type in our code.

**Comment by andreievg** [comment]
> Not sure what you mean- can you elaborate?

Why PluginArea, PluginType and Plugin<T>.module ? I think reducing the amount of things to consider when plugin is created is worth considering.

> I disagree about HMR and don't see that as a hard requirement. Besides which, hot reloading is working already in the examples I've done, although due to the convoluted loading nature, the whole page refreshes. Still, plugins are not intended as large applications and I don't expect that a hot reload is a big issue.

I think we've experience the difference that dev ops speed makes.  Oops I didn't quite go all the way up the PR list yet, so didn't quite see hot reloading working yet. When I was looking at current implementation I could only see it working by serving remote plugin via webpack server and accessing it that way (which once again requires more moving bits, with hot reloading and auto loading local plugins in the example)

> I don't like the approach of having specific plugin types though, having the types of available plugins hard coded and defined in code - I was trying to avoid that inflexibility.

I don't understand this, `inflexibility` here meaning API input being strongly typed ? 

>  At the plugin boundary we can only do limited plugin validation anyway and have to cast to the expected type at some point, i.e. trust that the plugin is implemented correctly. Anyway after this initial type cast we should use the correct type in our code.

The discriminated union approach mean that plugin type and plugin input are linked together

I would be keen to discuss it in a wider group.

**Comment by clemens-msupply** [comment]
> Why PluginArea, PluginType and Plugin.module ? I think reducing the amount of things to consider when plugin is created is worth considering.

This got me thinking a bit more. Instead of letting the plugin provide a list of components we could also follow the approach to have a `Plugin` init script where the plugin can access some plugin api/sdk to register components. For example: 

```ts
// plugin-main.ts

omSupplyPlugins.dashboard.registerWidget(() => (<MyDashboardWidget/>));
omSupplyPlugins.inboundShipment.toolbar.registerButton(...);
omSupplyPlugins.addEventListener(...);

// do other init things like
const myWorker = new Worker("plugin-worker.js");
```

This approach is actually relatively common for plugins. For example, browser extensions work like that.

Other question, do plugins have to use React? Ideally not, but might be too much extra work...

**Comment by mark-prins** [comment]
> I notice that potentially plugin is being loaded more then once ?

Not seeing that myself, will check the other examples as well tho

image

Thanks for the example @andreievg - that's very nice. Will integrate the webpack changes and see how we go. 
I was deliberately trying to avoid having the webpack config load plugins directly, as I still have in mind the concept of plugins built in isolation. Not so keen on requiring the full client app in order to build a plugin, but will head that way until I can come up with a good argument why not.

And I hear what you and @clemens-msupply are saying about stricter typing and naming. It means shifting things out of `common` and into the `host` app, I'm holding on to the intention of `common` having no app-specific code.

**Comment by mark-prins** [comment]
> Why PluginArea, PluginType and Plugin.module ?

PluginType you have in your example too - giving each plugin a type to specify the expected type of data and potentially page which it should appear on. 

The `PluginArea` defines where the component should render.
Looking at your example, the `OrderPlugin` can only add a react component into a single place within the Order page. Without the area being specified, we would require finer grained types such as the `OrderAppBarButtonPlugin`. There's a lot of crossover in the way our app is structured - most types have an AppBar and a Toolbar and DetailPanel etc. So it makes sense to lift the Area out and specify it directly.

The `module` is the actual react component, which you have as `Component`.

**Comment by andreievg** [comment]
> This got me thinking a bit more. Instead of letting the plugin provide a list of components we could also follow the approach to have a Plugin init script where the plugin can access some plugin api/sdk to register components. For example:

@clemens-msupply 

So we pass in omSupplyPlugins to the plugin, and the plugin adds whatever components/methods it implements to a bunch of collections/array that are provided through omSupplyPlugins ? On the first look, i like this, wonder if it's possible to let the host app know that all plugins have been `loaded` (or do we need to make the entire app wait before all plugins are loaded ?) and how to provide that `api/sdk to register components`. Although I think if api/sdk is provided as a global/singleton then it's a bit harder to figure out what is happening and not sure how it would work with federation module, the alternative is registerPlugin method provided by plugin that accepts omSupplyPlugins instance. 
I am not sure what the pros and cons of this is vs pluginProvider tbh.

@mark-prins 

> I was deliberately trying to avoid having the webpack config load plugins directly, as I still have in mind the concept of plugins built in isolation
> Not so keen on requiring the full client app in order to build a plugin

Personally I am still very strong on plugins being developer in omSupply dev environment. It seems faster, safer and easier to adjust if compatibility changes. It would be great to understand your preference though, hopefully we can get an alignment and closure on this soon.

> It means shifting things out of common and into the host app, I'm holding on to the intention of common having no app-specific code.

To be honest, I've been wondering if we need to revisit the drive toward the package splitting, I think it would only be beneficial if we end up sharing packages between projects/repos, and from my understanding, this is becoming more and more of a distant requirement, but it is adding a little bit of extra overhead (and some design constraints)

> Without the area being specified, we would require finer grained types such as the OrderAppBarButtonPlugin

I guess my whole questions comes down to this. Why not have OrderAppBarButtonPlugin ? Btw my concern for splitting up plugin type to different categories is that it would be hard to navigate to where this plugin is actually used, I think click through or find all type of codebase navigation is super important, I found it really hard to navigate to actual string column definition or through a bunch of files to API implementation. I think single type category would be easier to find and easier to use in a plugin (less config).

We can still add some structure to the type, like so:
```typescript

export type Plugins = {
  Order: {
    AppBar: {
      Button: {
        Component: React.ComponentType<{ data: { invoiceNode: Invoice } }>;
        pluginType: 'Plugins.AppBar.Button';
      };
    };
    DetailView: {
      Button: {
        Component: React.ComponentType<{ data: { invoiceNode: Invoice } }>;
        pluginType: 'Plugins.DetailView.Button';
      };
    };
  };
};

type PluginType = Plugins['Order']['AppBar']['Button'] | Plugins['Order']['DetailView']['Button'];\

const ShowInvoiceStatus: Plugins['Order']['AppBar']['Button'] = {
  Component: ({ data: { invoiceNode } }) => <div>status: {invoiceNode.status}</div>,
  pluginType: 'Plugins.AppBar.Button',
  name: 'ShowInvoiceStatus',
};
```

> The module is the actual react component, which you have as Component.

Yep, thanks, I found what it refers too, but was wondering why not just 'plugin' as per: 

In [loadPlugin](https://github.com/andreievg/plugins-example/blob/d009b355d5d4cfed18a858b393967a310fec8b47/frontend/src/plugin.tsx#L161) and in [plugin's webpack config](https://github.com/andreievg/plugins-example/blob/d009b355d5d4cfed18a858b393967a310fec8b47/frontend/src/plugins/First/webpack.config.js#L40).

> Not seeing that myself, will check the other examples as well tho

Your screenshot shows plugins served by webpack, can you please try with plugin served by back end ?

Btw, I think what you've done so far look very good and would work nicely. My suggestion are only improvements, I do think they are worth considering, but not necessary.

**Comment by clemens-msupply** [comment]
> > This got me thinking a bit more. Instead of letting the plugin provide a list of components we could also follow the approach to have a Plugin init script where the plugin can access some plugin api/sdk to register components. For example:
> 
> @clemens-msupply
> 
> So we pass in omSupplyPlugins to the plugin, and the plugin adds whatever components/methods it implements to a bunch of collections/array that are provided through omSupplyPlugins ? On the first look, i like this, wonder if it's possible to let the host app know that all plugins have been `loaded` (or do we need to make the entire app wait before all plugins are loaded ?) and how to provide that `api/sdk to register components`. Although I think if api/sdk is provided as a global/singleton then it's a bit harder to figure out what is happening and not sure how it would work with federation module, the alternative is registerPlugin method provided by plugin that accepts omSupplyPlugins instance. I am not sure what the pros and cons of this is vs pluginProvider tbh.

Yes think the best would be to let the host load/manage all the plugins and call an `init(omSupplyPlugins: OmSupplyPluginAPI)` method which is implemented by each plugin. The host can then keep track what plugin have been loaded.

>  I am not sure what the pros and cons of this is vs pluginProvider tbh.
Think the pro would be that you have clear lifecycle, e.g. you have a method that is called exactly once. Currently, I think you would need to load a component first before you could start something like a plugin background service?

**Pull request merged**